File: Solutions/188. Best Time to Buy and Sell Stock IV/CPP/Best_Time_to_Buy_and_Sell_Stock_IV.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution{
public:
    int maxProfit(int k, vector<int> &prices){
        vector<int> profitAfter_b(k + 1, INT_MIN);
        vector<int> profitAfter_s(k + 1, 0);

        //greedy algorithm
        int prices_size = prices.size();
        for(int i = 0; i < prices_size; i++){
            for(int j = 1; j <= k; j++){
                //iteratively look for the best buyTime[j] and sellTime[j] (from prices[0] ~ prices[i])
                profitAfter_b[j] = max(profitAfter_b[j], profitAfter_s[j - 1] - prices[i]);
                profitAfter_s[j] = max(profitAfter_s[j], profitAfter_b[j]     + prices[i]);
            }
        }

        return profitAfter_s[k];
    }
};

int main(){
    Solution* S = new Solution();

    //inputs
    int k = 2;
    vector<int> prices{3, 2, 6, 5, 0, 3};

    cout << S->maxProfit(k, prices) << endl;

    return 0;
}

File: Solutions/188. Best Time to Buy and Sell Stock IV/Python/Best_Time_to_Buy_and_Sell_Stock_IV.py
class Solution:
    def maxProfit(self, k: int, prices: list[int]) -> int:
        profitAfter_b = [float('-inf')] * (k + 1)
        profitAfter_s = [0] * (k + 1)

        for p in prices:
            for i in range(1, k + 1):
                profitAfter_b[i] = max(profitAfter_b[i], profitAfter_s[i - 1] - p)
                profitAfter_s[i] = max(profitAfter_s[i], profitAfter_b[i]     + p)

        return profitAfter_s[k]

S = Solution()

#inputs
k = 2
prices = [3, 2, 6, 5, 0, 3]

print(S.maxProfit(k, prices))